File: frontend/src/types/replay.ts
export type ReplayStatus = 'success' | 'failed' | 'partial' | 'unknown'

export type ReplayCounts = {
  messages: number
  tool_calls: number
  skills_used: number
  files_changed?: number | null
  tests_passed?: number | null
  tests_failed?: number | null
  subagents?: number | null
}

export type ReplayHashes = {
  source_hash?: string | null
  redacted_replay_hash?: string | null
  receipt_hash?: string | null
}

export type ReplayRun = {
  replay_id: string
  source_session_id: string
  title: string
  status: ReplayStatus
  started_at?: string | null
  ended_at?: string | null
  duration_ms?: number | null
  primary_model?: string | null
  total_cost_usd?: number | null
  counts: ReplayCounts
  hashes: ReplayHashes
  redaction_status: string
  created_at?: string | null
  updated_at?: string | null
}

export type ReplayEvent = {
  event_id: string
  replay_id: string
  type: string
  title: string
  summary: string
  raw_content?: string | null
  redacted_content?: string | null
  timestamp?: string | null
  duration_ms?: number | null
  status: string
  metadata?: Record<string, unknown>
}

export type RunReceipt = {
  schema_version: string
  receipt_id: string
  replay_id: string
  source_session_id: string
  title: string
  status: string
  started_at?: string | null
  ended_at?: string | null
  duration_ms?: number | null
  model?: string | null
  total_cost_usd?: number | null
  tool_call_count: number
  skills_used: Array<{ name: string; version?: string | null; hash?: string | null }>
  files_changed?: number | null
  hashes: ReplayHashes
  redaction: {
    mode: string
    findings_count: number
    redacted_fields_count: number
  }
  generated_at: string
  generator: {
    name: string
    version: string
  }
}

export type ReplayArtifact = {
  artifact_id: string
  replay_id: string
  type: string
  title: string
  summary?: string | null
  event_id?: string | null
  path?: string | null
  content?: string | null
  redacted_content?: string | null
  mime_type?: string | null
  size_bytes?: number | null
  hash?: string | null
}

export type ReplayDetail = {
  run: ReplayRun
  events: ReplayEvent[]
  artifacts: ReplayArtifact[]
  receipt?: RunReceipt | null
  missing_data: string[]
  redactions?: Array<{
    finding_id: string
    severity: string
    type: string
    field_path: string
    preview: string
    replacement: string
    auto_redacted: boolean
  }>
}

export type ReplayRunsResponse = {
  runs: ReplayRun[]
}

File: backend/models/replay.py
"""Replay data models."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal


ReplayStatus = Literal["success", "failed", "partial", "unknown"]
ReplayEventType = Literal[
    "prompt",
    "assistant_message",
    "tool_call",
    "tool_result",
    "terminal_command",
    "test_passed",
    "test_failed",
    "error",
    "completion",
]
EventStatus = Literal["success", "failed", "warning", "info", "unknown"]


@dataclass
class ReplayCounts:
    messages: int = 0
    tool_calls: int = 0
    skills_used: int = 0
    files_changed: int | None = None
    tests_passed: int | None = None
    tests_failed: int | None = None
    subagents: int | None = None


@dataclass
class ReplayHashes:
    source_hash: str | None = None
    redacted_replay_hash: str | None = None
    receipt_hash: str | None = None


@dataclass
class ReplayRun:
    replay_id: str
    source_session_id: str
    title: str
    status: ReplayStatus
    counts: ReplayCounts
    hashes: ReplayHashes = field(default_factory=ReplayHashes)
    redaction_status: str = "not_scanned"
    started_at: str | None = None
    ended_at: str | None = None
    duration_ms: int | None = None
    primary_model: str | None = None
    total_cost_usd: float | None = None
    created_at: str | None = None
    updated_at: str | None = None


@dataclass
class ReplayEvent:
    event_id: str
    replay_id: str
    type: ReplayEventType
    title: str
    summary: str
    status: EventStatus
    raw_content: str | None = None
    redacted_content: str | None = None
    timestamp: str | None = None
    duration_ms: int | None = None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class ReplayArtifact:
    artifact_id: str
    replay_id: str
    type: str
    title: str
    summary: str | None = None
    event_id: str | None = None
    path: str | None = None
    content: str | None = None
    redacted_content: str | None = None
    mime_type: str | None = None
    size_bytes: int | None = None
    hash: str | None = None


@dataclass
class RedactionFinding:
    finding_id: str
    severity: Literal["low", "medium", "high", "critical"]
    type: str
    field_path: str
    preview: str
    replacement: str
    auto_redacted: bool = True


@dataclass
class RunReceipt:
    schema_version: str
    receipt_id: str
    replay_id: str
    source_session_id: str
    title: str
    status: str
    tool_call_count: int
    skills_used: list[dict[str, str | None]]
    hashes: ReplayHashes
    redaction: dict[str, str | int]
    generated_at: str
    generator: dict[str, str]
    started_at: str | None = None
    ended_at: str | None = None
    duration_ms: int | None = None
    model: str | None = None
    total_cost_usd: float | None = None
    files_changed: int | None = None
    tests: dict[str, int | str | None] | None = None


@dataclass
class ReplayDetail:
    run: ReplayRun
    events: list[ReplayEvent] = field(default_factory=list)
    artifacts: list[ReplayArtifact] = field(default_factory=list)
    receipt: RunReceipt | None = None
    missing_data: list[str] = field(default_factory=list)
    redactions: list[RedactionFinding] = field(default_factory=list)


@dataclass
class ReplayExportResult:
    ok: bool
    export_path: str
    receipt_hash: str | None
    redacted_replay_hash: str | None
    redaction_notice: str
